### PR TITLE
feat(sink): Support without backfill sink with 'create sink from mv'.

### PR DESCRIPTION
**Pull request by xxhZs**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Support without backfill sink with 'create sink from mv'.(Don't support with create sink as query).

We can use `create sink from mv with(without_backfill = true).`  to create a sink without backfill.

The syntax regarding 'without backfill' may need further discussion.

issue: https://github.com/risingwavelabs/risingwave/issues/13102

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by kwannoel** [comment]
We can generalize it to `with (chain_type = 'upstream_only')`.

**Comment by hzxa21** [comment]
Is this PR ready for review?

**Comment by xxhZs** [comment]
> Is this PR ready for review?

If there is no dispute, I will resolve the conflict for review

**Comment by hzxa21** [comment]
> > Is this PR ready for review?
> 
> If there is no dispute, I will resolve the conflict for review

I think it is okay to support without snapshot only for `CREATE SINK FROM` but not for `CREATE SINK AS`.

cc @BugenZhao @fuyufjh @neverchanje Any thoughts?

File: src/connector/src/sink/mod.rs [hunk 1]
```diff
@@ -132,6 +132,7 @@ macro_rules! match_sink_name_str {
 
 pub const CONNECTOR_TYPE_KEY: &str = "connector";
 pub const SINK_TYPE_OPTION: &str = "type";
+pub const SINK_WITHOUT_BACKFILL: &str = "without_backfill";
```
Comment (hzxa21) [inline src/connector/src/sink/mod.rs:135]: Personally I perfer calling it `incremental` instead because user may not understand what "backfill" is.

**Review by hzxa21 — APPROVED** [review]
LGTM. Can you add some tests for this feature?

**Comment by hzxa21** [comment]
@kwannoel @chenzl25 PTAL

File: src/connector/src/sink/mod.rs (hunk 1 above)
Comment (BugenZhao) [reply src/connector/src/sink/mod.rs:135]: Agree. `without` is a privative and `backfill` is a terminology of internal implementation. What about simply calling it `snapshot`?

File: src/frontend/src/optimizer/mod.rs [hunk 2]
```diff
@@ -717,9 +733,16 @@ impl PlanRoot {
         db_name: String,
         sink_from_table_name: String,
         format_desc: Option<SinkFormatDesc>,
+        without_backfail: bool,
```
Comment (BugenZhao) [inline src/frontend/src/optimizer/mod.rs:736]:
```suggestion
        without_backfill: bool,
```

File: src/frontend/src/optimizer/plan_node/convert.rs [hunk 3]
```diff
@@ -124,16 +125,29 @@ impl RewriteStreamContext {
...
+        stream_scan_type: StreamScanType,
+    ) -> Self {
         Self {
             share_to_stream_map: HashMap::new(),
             emit_on_window_close,
+            stream_scan_type,
         }
     }
 
+    pub fn get_stream_scan_type(&self) -> StreamScanType {
```
Comment (BugenZhao) [inline src/frontend/src/optimizer/plan_node/convert.rs:147]:
```suggestion
    pub fn stream_scan_type(&self) -> StreamScanType {
```

**Review by BugenZhao — COMMENTED** [review]
Rest LGTM

**Comment by kwannoel** [comment]
Rest LGTM. Could you also add some e2e tests for this feature, mark this PR as user-facing and update the release notes?

File: src/frontend/src/handler/create_sink.rs [hunk 4]
```diff
@@ -105,13 +105,16 @@ pub fn gen_sink_plan(
 
     // Used for debezium's table name
     let sink_from_table_name;
+    let sink_form_is_from;
```
[the remark this answers is not in the document]
Comment (chenzl25) [reply src/frontend/src/handler/create_sink.rs:108]: `sink_form_is_from` looks like a typo at first glance. How about we name it `direct_sink`?

File: src/frontend/src/handler/create_sink.rs [hunk 5]
```diff
@@ -190,6 +193,20 @@ pub fn gen_sink_plan(
         plan_root.set_out_names(col_names)?;
     };
 
+    let without_backfill = match with_options.get(SINK_WITHOUT_BACKFILL) {
+        Some(flag) if flag.eq("true") => {
```
Comment (tabVersion) [inline src/frontend/src/handler/create_sink.rs:197]: ignore case?

**Review by tabVersion — APPROVED** [review]
LGTM, but still need a e2e test

**Pull request merged**